### PR TITLE
[LLVM-Reduce] - Distinct Metadata Reduction

### PR DESCRIPTION
**Pull request by rbintel**
* Add a distinct metadata reduction pass, which traverses the whole unnamed metadata tree and applies reduction where possible. Previous version could do this only partially, either removing named metadata, metadata attached to instructions or debug information.

* Modify current named node operand reduction, make it more aggressive by generalizing the algorithm instead of reducing hard-coded instructions, I see no issue in trying a more aggressive reduction and rolling it back in case it doesn't go through.

* Refactor some of the tests to suit new functionality:

- Remove the --abort-on-invalid-reduction flag from remove-dp-values.ll tests, if it is included, the new named metadata reduction algorithm will fail at some point, if not, the test passes, valid IR is generated and the module is reduced, which costs a few more iterations.

- Refactor remove-metadata.ll, the new functionality will not only erase the top level nodes but also their children nodes, so one can't expect them to be present after the run.

- Refactor remove-named-metadata.ll, the expected behaviour now is to also remove the operands of !some.unknown.named.

- Add a test for the new functionality, expected behaviour is to leave no boring nodes and all interesting nodes.

### CODE REVIEW

File: llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    }
+  }
+  for (NamedMDNode &namedNode : Program.named_metadata())
+    cleanup(namedNode, tmp, Program);
+}
+
+void llvm::reduceDistinctMetadataDeltaPass(TestRunner &Test) {
+  runDeltaPass(Test, extractDistinctMetadataFromModule,
+               "Reducing Distinct Metadata");
+}
```
Comment (michalpaszkowski) [inline llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp:146]: Please add new lines at the end of files

File: llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp [hunk 2]
```diff
@@ -0,0 +1,146 @@
+//===- ReduceDistinctMetadata.cpp - Specialized Delta Pass
```
Comment (michalpaszkowski) [inline llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp:1]: Please fix the comment

File: llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.h [hunk 3]
```diff
@@ -0,0 +1,24 @@
+//===- ReduceDistinctMetadata.h - Specialized Delta Pass
```
Comment (michalpaszkowski) [inline llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.h:1]: Please fix the comment

File: llvm/test/tools/llvm-reduce/remove-dp-values.ll [hunk 4]
```diff
@@ -1,8 +1,8 @@
-; RUN: llvm-reduce --abort-on-invalid-reduction --test FileCheck --test-arg --check-prefixes=CHECK-INTERESTINGNESS --test-arg %s --test-arg --input-file %s -o %t
+; RUN: llvm-reduce --test FileCheck --test-arg --check-prefixes=CHECK-INTERESTINGNESS --test-arg %s --test-arg --input-file %s -o %t
```
Comment (michalpaszkowski) [inline llvm/test/tools/llvm-reduce/remove-dp-values.ll:1]: Could you please clarify why the flag `--abort-on-invalid-reduction` needs to be removed? I see your point in the PR description, but I am not sure I understand correctly.

File: llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+using namespace llvm;
+
+// Traverse the graph breadth-first and try to remove unnamed metadata nodes
+void BFS_removal(MDNode *root,
+                 std::set<std::pair<unsigned int, MDNode *>> &nodesToDelete,
+                 MDNode *tmp, Oracle &O, Module &Program) {
+  std::queue<MDNode *> q{};
+  std::set<MDNode *>
+      visited{}; // Keep track of visited nodes not to get into loops
```
Comment (michalpaszkowski) [inline llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp:32]: Please follow the [LLVM Coding Standards](https://llvm.org/docs/CodingStandards.html#id43), e.g. when it comes to variable names.

**Review by michalpaszkowski — CHANGES_REQUESTED** [review]
Hi @rbintel! Thank you for the pull request. I added some initial comments on the style below.

File: llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+  while (!q.empty()) {
+    MDNode *current = q.front();
+    q.pop();
+
+    // Mark the nodes for removal
+    for (unsigned int i = 0; i < current->getNumOperands(); ++i) {
+      Metadata *operand = current->getOperand(i).get();
+      if (isa_and_nonnull<MDNode>(operand)) {
+        if (std::find(visited.begin(), visited.end(), operand) ==
```
Comment (michalpaszkowski) [inline llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp:43]: Using std::find on std::set might be inefficient. You could use visited.find(operand) != visited.end().

File: llvm/test/tools/llvm-reduce/remove-dp-values.ll (hunk 4 above)
Comment (rbintel) [reply llvm/test/tools/llvm-reduce/remove-dp-values.ll:1]: So the problem is I changed the way first level unnamed metadata reduction is done (ReduceMetadata.cpp). At some point it will try to reduce the nodes in `!llvm.dbg.cu`, so the reducer will generate an invalid case and go back if there is no flag (or fail if the flag is set). This happens due to the fact that I removed the hardcoded constraints, before it was pretty conservative and only tried to run the algorithm on a subset of named metadata:
```
static constexpr StringLiteral ListNamedMetadata[] = {
  "llvm.module.flags",
  "llvm.ident",
  "opencl.spir.version",
  "opencl.ocl.version",
  "opencl.used.extensions",
  "opencl.used.optional.core.features",
  "opencl.compiler.options"
};
```
but now it tries to reduce any named node it encounters.

The question is which behavior is preferable:
1) Try the reduction, fail the test and continue reducing
2) Skip the named node if it is `!llvm.dbg.cu`

**Review by michalpaszkowski — APPROVED** [review]
Thank you! I think the change looks good to me! One last small ask, please update the files which are missing new lines at the end. I think for .ll files this is a good practice to add a new line, but for Python files missing a new line is a PEP8 error.

**Comment by michalpaszkowski** [comment]
@SLTozer @OCHyams @ormris @arsenm I am tagging you all in case you have anything to add in the review. Would be grateful for any additional comments as you have more experience in this area.

**Comment by arsenm** [comment]
> Remove the --abort-on-invalid-reduction flag

We should only ever be adding uses of it, not removing it. We should always be trying to avoid invalid IR

File: llvm/test/tools/llvm-reduce/Inputs/remove-metadata.py [hunk 7]
```diff
@@ -5,4 +5,4 @@
   if "!interesting" in line:
     sys.exit(0)
 
-sys.exit(1)
+sys.exit(1)
```
Comment (arsenm) [inline llvm/test/tools/llvm-reduce/Inputs/remove-metadata.py:8]: Missing newline end of file error

File: llvm/test/tools/llvm-reduce/reduce-distinct-metadata.ll [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+!4 = distinct !{!"interesting_1", !6, i2 1}
+!5 = distinct !{!"interesting_2", !8}
+!6 = distinct !{!"interesting_2", !10}
+!7 = distinct !{!"interesting_2", !12}
+!8 = distinct !{!"interesting_3", !10, !9}
+!9 = distinct !{!"interesting_3", !11, !13}
+!10 = distinct !{!"boring_1", i32 50}
+!11 = distinct !{!"boring_1", i32 2}
+!12 = distinct !{!"boring_3", i2 1}
+!13 = distinct !{!"interesting_4"}
```
Comment (arsenm) [inline llvm/test/tools/llvm-reduce/reduce-distinct-metadata.ll:28]: Missing end of file error r

File: llvm/test/tools/llvm-reduce/remove-dp-values.ll (hunk 4 above)
Comment (arsenm) [reply llvm/test/tools/llvm-reduce/remove-dp-values.ll:1]: The main reason this checks for understood metadata is to avoid hitting verifier errors in out of tree verifiers. We need to know the expected structure of the node to not break whatever other code is expecting

File: llvm/test/tools/llvm-reduce/remove-dp-values.ll [hunk 9]
```diff
@@ -1,8 +1,8 @@
...
 ; RUN: FileCheck --check-prefixes=CHECK-FINAL --input-file=%t %s --implicit-check-not=#dbg_value
 
 ; RUN: opt < %s -S --write-experimental-debuginfo=false > %t.intrinsics.ll
-; RUN: llvm-reduce --abort-on-invalid-reduction --test FileCheck --test-arg --check-prefixes=INTRINSIC-INTERESTINGNESS --test-arg %s --test-arg --input-file %t.intrinsics.ll -o %t
```
Comment (arsenm) [inline llvm/test/tools/llvm-reduce/remove-dp-values.ll:5]: abort-on-invalid-reduction should not be removed. It was mostly a hack for bringup, but we should try harder to avoid breaking the IR in the first place

File: llvm/test/tools/llvm-reduce/remove-metadata.ll [hunk 10]
```diff
@@ -2,7 +2,7 @@
 ; The Metadata pass erases named & unnamed metadata nodes.
 ;
 ; RUN: llvm-reduce --test %python --test-arg %p/Inputs/remove-metadata.py %s -o %t
-; RUN: cat %t | FileCheck -implicit-check-not=! %s
+; RUN: FileCheck -implicit-check-not=! %s < %t
```
Comment (arsenm) [inline llvm/test/tools/llvm-reduce/remove-metadata.ll:5]: Can precommit this unrelated change

File: llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+      Metadata *Operand = CurrentNode->getOperand(I).get();
+      if (isa_and_nonnull<MDNode>(Operand)) {
+        // Check whether node has been visited
+        if (VisitedNodes.find(static_cast<MDNode *>(Operand)) ==
+            VisitedNodes.end()) {
+          NodesToTraverse.push(static_cast<MDNode *>(Operand));
+          VisitedNodes.insert(static_cast<MDNode *>(Operand));
+        }
+        // Delete the node only if it is distinct
+        if (static_cast<MDNode *>(Operand)->isDistinct())
```
Comment (arsenm) [inline llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp:49]: Should never use static_cast, and use the llvm cast operation. Also, avoid repeating the cast so many times.

File: llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+  NodesToTraverse.push(Root);
+
+  while (!NodesToTraverse.empty()) {
+    MDNode *CurrentNode = NodesToTraverse.front();
+    NodesToTraverse.pop();
+
+    // Mark the nodes for removal
+    for (unsigned int I = 0; I < CurrentNode->getNumOperands(); ++I) {
+      Metadata *Operand = CurrentNode->getOperand(I).get();
+      if (isa_and_nonnull<MDNode>(Operand)) {
```
Comment (arsenm) [inline llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp:41]: dyn_cast here instead of this isa and static_cast check

File: llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+
+using namespace llvm;
+
+// Traverse the graph breadth-first and try to remove unnamed metadata nodes
+void reduceNodes(MDNode *Root,
+                 std::set<std::pair<unsigned int, MDNode *>> &NodesToDelete,
+                 MDNode *TemporaryNode, Oracle &O, Module &Program) {
+  std::queue<MDNode *> NodesToTraverse{};
+  // Keep track of visited nodes not to get into loops
+  std::set<MDNode *> VisitedNodes{};
```
Comment (arsenm) [inline llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp:31]: Can you use a SetVector for this? Or at least use DenseSet instead of std::set?

File: llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+        }
+        // Delete the node only if it is distinct
+        if (static_cast<MDNode *>(Operand)->isDistinct())
+          // Add to removal list
+          NodesToDelete.insert(std::make_pair(I, CurrentNode));
+      }
+    }
+
+    // Remove the nodes
+    for (std::pair<unsigned int, MDNode *> Node : NodesToDelete) {
```
Comment (arsenm) [inline llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp:56]: Use C++17 destructuring?

File: llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+    NodesToDelete.clear();
+  }
+}
+
+// After reducing metadata, we need to remove references to the temporary node,
+// this is also done with BFS
+void cleanUpTemporaries(NamedMDNode &NamedNode, MDTuple *TemporaryTuple,
+                        Module &Program) {
+  std::queue<MDTuple *> NodesToTraverse{};
+  std::set<MDTuple *> VisitedNodes{};
```
Comment (arsenm) [inline llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp:69]: Same as above, avoid std::set

File: llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+
+  // Push all first level operands of the named node to the queue
+  for (auto I = NamedNode.op_begin(); I != NamedNode.op_end(); ++I) {
+    // If the node hasn't been traversed yet, add it to the queue of nodes to
+    // traverse.
+    if (VisitedNodes.find(static_cast<MDTuple *>(*I)) == VisitedNodes.end()) {
```
Comment (arsenm) [inline llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp:75]: Avoid static_cast

File: llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+  }
+}
+
+static void extractDistinctMetadataFromModule(Oracle &O,
+                                              ReducerWorkItem &WorkItem) {
+  Module &Program = WorkItem.getModule();
+  MDTuple *TemporaryTuple = MDTuple::getDistinct(
+      Program.getContext(), SmallVector<Metadata *, 1>{llvm::MDString::get(
+                                Program.getContext(), "temporary_tuple")});
+  std::set<std::pair<unsigned int, MDNode *>> NodesToDelete{};
```
Comment (arsenm) [inline llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp:132]: DenseSet or SmallPtrSet

File: llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+  for (NamedMDNode &NamedNode :
+       Program.named_metadata()) { // Iterate over the named nodes
+    for (unsigned int I = 0; I < NamedNode.getNumOperands();
+         ++I) { // Iterate over first level unnamed nodes..
+      Metadata *Operand = NamedNode.getOperand(I);
+      if (isa_and_nonnull<MDTuple>(Operand))
```
Comment (arsenm) [inline llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp:138]: dyn_cast

File: llvm/tools/llvm-reduce/deltas/ReduceMetadata.cpp [hunk 19]
```diff
@@ -28,40 +28,24 @@ static bool shouldKeepDebugNamedMetadata(NamedMDNode &MD) {
   return MD.getName() == "llvm.dbg.cu" && MD.getNumOperands() != 0;
 }
 
-// Named metadata with simple list-like behavior, so that it's valid to remove
-// operands individually.
```
Comment (arsenm) [inline llvm/tools/llvm-reduce/deltas/ReduceMetadata.cpp:32]: We should preserve this. If we want a mode where we blindly rip operands out of anything we don't understand, probably should be a flag

**Comment by OCHyams** [comment]
> @SLTozer @OCHyams @ormris @arsenm I am tagging you all in case you have anything to add in the review. Would be grateful for any additional comments as you have more experience in this area.

Sorry for the radio-silence on this, I've been on and off PTO and this slipped through the cracks for me

**Comment by arsenm** [comment]
What's going on with this? The last batch of comments weren't addressed and it was closed without comment?

**Comment by rbintel** [comment]
@arsenm Closed this pull request because of a bad rebase. Will reopen another one with code review fixes and a clean commit history.
